pkg/core/bundle_unpack.go

```diff
@@ -60,72 +164,305 @@ type errorHit struct {
+	bundle *Bundle,
+	fs cafs.Fs,
+	chans downloadBundleChans) {
+	defer func() {
+		<-chans.concurrencyControl
+	}()
+	reportError := func(err error) {
+		select {
+		case chans.error <- errorHit{
+			err,
+			bundleEntry.NameWithPath,
+		}:
+		case <-chans.done:
+		}
+	}
+	err := downloadBundleEntrySyncMaybeOverwrite(ctx, bundleEntry, bundle, fs, true)
+	if err != nil {
+		reportError(err)
+		return
+	}
+}
+
+func deleteBundleEntry(ctx context.Context, bundleEntry model.BundleEntry,
+	bundle *Bundle,
+	chans downloadBundleChans) {
+	defer func() {
+		<-chans.concurrencyControl
+	}()
+	reportError := func(err error) {
+		select {
+		case chans.error <- errorHit{
+			err,
+			bundleEntry.NameWithPath,
+		}:
+		case <-chans.done:
+		}
+	}
+	err := deleteBundleEntrySync(ctx, bundleEntry, bundle)
+	if err != nil {
+		reportError(err)
+		return
+	}
+}
+
+func downloadBundleEntries(ctx context.Context,
+	bundle *Bundle,
+	bundleDest *Bundle,
+	diff BundleDiff,
+	fs cafs.Fs,
+	chans downloadBundleChans) {
+
+	reportError := func(err error) {
+		select {
+		case chans.error <- errorHit{
+			err,
+			"",
+		}:
+		case <-chans.done:
+		}
+	}
+
 	concurrentFileDownloads := bundle.concurrentFileDownloads
 	if concurrentFileDownloads < 1 {
 		concurrentFileDownloads = 1
 	}
 	concurrencyControl := make(chan struct{}, concurrentFileDownloads)
-	fmt.Printf("Downloading %d files\n", len(bundle.BundleEntries))
-	for _, b := range bundle.BundleEntries {
-		if file != "" && file != b.NameWithPath {
-			continue
+	chans.concurrencyControl = concurrencyControl
+	if bundleDest == nil {
+		bundle.l.Info("downloading bundle entries",
+			zap.Int("num", len(bundle.BundleEntries)))
+		for _, b := range bundle.BundleEntries {
+			concurrencyControl <- struct{}{}
+			go downloadBundleEntry(ctx, b, bundle, fs, chans)
 		}
-		wg.Add(1)
-		go func(bundleEntry model.BundleEntry, cc chan struct{}) {
-			cc <- struct{}{}
-			defer func() {
-				<-cc
-			}()
-
-			defer wg.Done()
-			fmt.Println("started " + bundleEntry.NameWithPath)
-			key, err := cafs.KeyFromString(bundleEntry.Hash)
-			if err != nil {
-				errC <- errorHit{
-					err,
-					bundleEntry.NameWithPath,
-				}
-				return
-			}
-			reader, err := fs.Get(ctx, key)
-			if err != nil {
-				errC <- errorHit{
-					err,
-					bundleEntry.NameWithPath,
-				}
-				return
-			}
-			err = bundle.ConsumableStore.Put(ctx, bundleEntry.NameWithPath, reader, storage.IfNotPresent)
-			if err != nil {
-				fmt.Printf("Failed to download %s error %s", bundleEntry.NameWithPath, err)
-				errC <- errorHit{
-					err,
-					bundleEntry.NameWithPath,
-				}
-				return
+	} else {
+		bundle.l.Info("downloading diff entries",
+			zap.Int("num", len(diff.Entries)))
+		for _, de := range diff.Entries {
+			concurrencyControl <- struct{}{}
+			switch de.Type {
+			case DiffEntryTypeAdd:
+				bundle.l.Info("adding added entry",
+					zap.String("name Additional", de.Additional.NameWithPath),
```

Comment:
throughout this `switch`, `de.Name` can replace `de.Additional.NameWithPath` and `de.Existing.NameWithPath`